app/api/api/filters.py

```diff
@@ -0,0 +1,148 @@
+from django.db.models.query_utils import Q
+from rest_framework import filters
+from shared.data.models import VisibilityChoices
+
+
+class ScanReportAccessFilter(filters.BaseFilterBackend):
+    """
+    Filter that only allows users to see Scan Reports they are allowed to view, edit, or admin.
+
+    """
+
+    # Each model type needs a different relationship to get the Scan Report / Dataset permissions.
+    RELATIONSHIP_MAPPING = {
+        "scanreport": "",
+        "scanreporttable": "scan_report__",
+        "scanreportfield": "scan_report_table__scan_report__",
+        "scanreportvalue": "scan_report_field__scan_report_table__scan_report__",
+    }
+
+    def filter_queryset(self, request, queryset, view):
+        """
+        Filters the queryset based on visibility and permissions related to datasets and scan reports.
+
+        Args:
+            - request: The request object.
+            - queryset: The queryset to be filtered based on visibility and permissions.
+
+        Returns:
+            - A filtered queryset based on the specified visibility and permission conditions.
+        """
+
+        model = queryset.model.__name__.lower()
+        relationship = self.RELATIONSHIP_MAPPING.get(model, "")
+
+        dataset_visibility = f"{relationship}parent_dataset__visibility"
+        scan_report_visibility = f"{relationship}visibility"
+        scan_report_viewers = f"{relationship}viewers"
+        scan_report_editors = f"{relationship}editors"
+        scan_report_author = f"{relationship}author"
+        dataset_viewers = f"{relationship}parent_dataset__viewers"
+        dataset_editors = f"{relationship}parent_dataset__editors"
+        dataset_admins = f"{relationship}parent_dataset__admins"
+        project_members = f"{relationship}parent_dataset__project__members"
```

Comment:
I'm going to assume that you've checked this works, remembering that datasets to projects is many-to-many.

Reply:
Yep it works, and with multiple users involved (as in, multiple users being in the lists) - I was expecting to use `__in` (or however Django does it) but it wasn't necessary.